datafusion-examples/examples/udaf_expr.rs

```diff
@@ -0,0 +1,45 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use datafusion::{
+    execution::{
+        config::SessionConfig,
+        context::{SessionContext, SessionState},
+    },
+    functions_aggregate::{expr_fn::first_value, register_all},
+};
+
+use datafusion_common::Result;
+use datafusion_expr::col;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let ctx = SessionContext::new();
+    let config = SessionConfig::new();
+    let mut state = SessionState::new_with_config_rt(config, ctx.runtime_env());
+    let _ = register_all(&mut state);
+
+    let first_value_udaf = state.aggregate_functions().get("FIRST_VALUE").unwrap();
+    let first_value_builder = first_value_udaf
```

Comment:
The downside (?) of this version is that it needs an internal function (non-public), so it is not so trivial to use it, and `call` always expects vec of expr now.

But, there is no builder function like `first_value_builder`, so it maybe make less noise to the user (?).

Reply:
I'm thinking maybe we change the existing expr::fn to builder mode too.

There are two ways to build Expr
`first_value(arg).order_by.build()`
and
`first_value_uda.call().order_by.build()`

I suggest we force to use builder mode for aggregate and window function given there are many arguments to build Expr. Leave scalar function as it is. @alamb What do you think?